lib/Conversion/ArcToLLVM/LowerArcToLLVM.cpp

```diff
@@ -173,6 +173,68 @@ struct StorageGetOpLowering : public OpConversionPattern<arc::StorageGetOp> {
   }
 };
 
+struct AssertOpLowering : public OpConversionPattern<arc::AssertOp> {
+  using OpConversionPattern::OpConversionPattern;
+  LogicalResult
+  matchAndRewrite(arc::AssertOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const final {
+
+    Value negateCondition = rewriter.create<arith::XOrIOp>(
+        op.getLoc(), adaptor.getProperty(),
+        rewriter.create<arith::ConstantIntOp>(op.getLoc(), 1, 1));
+
+    rewriter.replaceOpWithNewOp<scf::IfOp>(
+        op, negateCondition,
+        [&](OpBuilder &builder, Location loc) {
+          // If the condition doesn't hold then print the message and abort
+          auto message = op.getMsg().value_or("None");
+          auto module = op->getParentOfType<ModuleOp>();
+
+          LLVM::LLVMFuncOp putsFunc;
+          LLVM::GlobalOp globalOp;
+
+          {
+            OpBuilder::InsertionGuard guard(rewriter);
+            rewriter.setInsertionPointToStart(module.getBody());
+            putsFunc = lookupOrCreateFn(
+                module, "puts", LLVM::LLVMPointerType::get(getContext()),
+                LLVM::LLVMVoidType::get(getContext()), false);
+            llvm::StringRef msgRef(message);
+
+            // Get the integer type with the correct bit width.
+            Type int8Type = IntegerType::get(op.getContext(), 8);
+
+            // Generate new names for global strings.
+            // Messgaes with the same contents are mapped to the same names
+            std::string msgGlobalName =
+                std::string("Arc_Assert_Op_Msg: ") + message.str();
```

Comment:
Maybe adding the source location to the message could be useful.

Reply:
I can't think of a way to get the location of the here are two cases:
- If the message is provided directly to the operation (`arc.assert bool, clock, "some message"`), so the location of it is the location of the operation (aka `arc.assert`), here we map the same message contents to the same global string(address) which is very useful.
- If the message is provided as a variable (`arc.assert bool, clock, StrOp`), I need to find a way to get the location of the operand and then map all users to the same message, as the location will be different for different messages, and maybe they have the same contents, but in different locations.

Is there any way to convert `Location` to `StringRef`?

Reply:
Yes, I mean the location that you get from `assertOp.getLoc()`. The `Location` class has a `print` function that you can use. Something like this:
```
std::string str;
llvm::raw_string_ostream(str);
loc.print(stream);
// use str here
```
Because the message is a string attribute, you can just concatenate them at compile-time and still merge globals referring to the same string.

This is not super important, so if you're having issues with it, we can also just leave it away.

Reply:
That's great, but I tried it now it generates a location like this:
```
/Path/To/circt/test/Conversion/ArcToLLVM/lower-arc-to-llvm.mlir:276:3

```
I can use some `regex` matching and the output will be `lower-arc-to-llvm.mlir:276:3`, but If someone tries to add new tests at the start of the file this will change the line of the assertion op, the test will fail.
can we add it without test coverage?

Reply:
That's great, but I tried it now it generates a location like this:
```
/Path/To/circt/test/Conversion/ArcToLLVM/lower-arc-to-llvm.mlir:276:3

```
I can use some `regex` matching and the output will be `lower-arc-to-llvm.mlir:276:3`,  but I think it will differ from path to path, I tried but I couldn't match it so I ignored this part!
If you have further suggestions, I will address them.